src/__tests__/waitForElement.js

```diff
@@ -0,0 +1,278 @@
+}
+
+async function skipSomeTimeForMutationObserver(delayMs = 50) {
+  // Using `setTimeout` >30ms instead of `wait` here because `mutationobserver-shim` uses `setTimeout` ~30ms.
+  await skipSomeTime(delayMs, 50)
+}
+
+test('it waits for the callback to return a value and only reacts to DOM mutations', async () => {
+  const {container, getByTestId} = render(
+    `<div data-testid="initial-element"></div>`,
+  )
+
+  let nextElIndex = 0
+  const makeMutationFn = () => () => {
+    container.appendChild(
+      render(
+        `<div data-testid="another-element-that-causes-mutation-${++nextElIndex}"></div>`,
+      ).container.firstChild,
+    )
+  }
+
+  const testEl = render(
+    `<div data-testid="the-element-we-are-looking-for"></div>`,
+  ).container.firstChild
+  testEl.parentNode.removeChild(testEl)
+
+  const mutationsAndCallbacks = [
+    [
+      makeMutationFn(),
+      () => {
+        throw new Error('First exception.')
+      },
+    ],
+    [
+      makeMutationFn(),
+      () => {
+        throw new Error('Second exception.')
+      },
+    ],
+    [makeMutationFn(), () => null],
+    [makeMutationFn(), () => undefined],
+    [makeMutationFn(), () => getByTestId('the-element-we-are-looking-for')],
+    [
+      () => container.appendChild(testEl),
+      () => getByTestId('the-element-we-are-looking-for'),
+    ],
+  ]
+
+  const callback = jest
+    .fn(() => {
+      throw new Error('No more calls are expected.')
+    })
+    .mockName('callback')
+    .mockImplementation(() => {
+      throw new Error(
+        'First callback call is synchronous, not returning any elements.',
+      )
+    })
+  const successHandler = jest.fn().mockName('successHandler')
+  const errorHandler = jest.fn().mockName('errorHandler')
+
+  const promise = waitForElement(callback, {container}).then(
+    successHandler,
+    errorHandler,
+  )
+
+  // One synchronous `callback` call is expected.
+  expect(callback).toHaveBeenCalledTimes(1)
+  expect(successHandler).toHaveBeenCalledTimes(0)
+  expect(errorHandler).toHaveBeenCalledTimes(0)
+
+  await skipSomeTimeForMutationObserver()
+
+  // No more expected calls without DOM mutations.
+  expect(callback).toHaveBeenCalledTimes(1)
+  expect(successHandler).toHaveBeenCalledTimes(0)
+  expect(errorHandler).toHaveBeenCalledTimes(0)
+
+  // Perform mutations one by one, waiting for each to trigger `MutationObserver`.
+  for (const [mutationImpl, callbackImpl] of mutationsAndCallbacks) {
+    callback.mockImplementation(callbackImpl)
+    mutationImpl()
+    await skipSomeTimeForMutationObserver() // eslint-disable-line no-await-in-loop
+  }
+
+  expect(callback).toHaveBeenCalledTimes(1 + mutationsAndCallbacks.length)
+  expect(successHandler).toHaveBeenCalledTimes(1)
+  expect(successHandler).toHaveBeenCalledWith(testEl)
+  expect(errorHandler).toHaveBeenCalledTimes(0)
+  expect(container).toMatchSnapshot()
+  expect(testEl.parentNode).toBe(container)
+
+  await promise
```

Comment:
Rather than `await promise`, I'd prefer `return promise`. Maybe I'm being irrational here, but it feels better to me.

Reply:
I'm changing this, but if this line was the only one with `await` expression, then `eslint` would say `async` function must have an `await`. We are lucky that there are more `await`s in all the tests.

P.S. I don't agree with this rule because `async` functions have their use for the error handling; `async` function is just the one that is wrapped in a `return new Promise(...)`.

Reply:
👉 https://github.com/kentcdodds/dom-testing-library/pull/10/commits/414a3dca4f75abb64033d7e614255b66e42bf694